pkg/controller/registry/resolver/solver/solve.go

```diff
@@ -0,0 +1,249 @@
+	// problem.
+	Identifier() Identifier
+	// Constraints returns the set of constraints that apply to
+	// this Installable.
+	Constraints() []Constraint
+}
+
+var Incomplete = errors.New("cancelled before a solution could be found")
+
+// NotSatisfiable is an error composed of a minimal set of applied
+// constraints that is sufficient to make a solution impossible.
+type NotSatisfiable []AppliedConstraint
+
+func (e NotSatisfiable) Error() string {
+	const msg = "constraints not satisfiable"
+	if len(e) == 0 {
+		return msg
+	}
+	s := make([]string, len(e))
+	for i, a := range e {
+		s[i] = a.String()
+	}
+	return fmt.Sprintf("%s: %s", msg, strings.Join(s, ", "))
+}
+
+type Solver interface {
+	Solve(context.Context) ([]Installable, error)
+}
+
+type solver struct {
+	dict   *dict
+	tracer Tracer
+	buffer []z.Lit
+}
+
+const (
+	satisfiable   = 1
+	unsatisfiable = -1
+)
+
+// Solve takes a slice containing all Installables and returns a slice
+// containing only those Installables that were selected for
+// installation. If no solution is possible, or if the provided
+// Context times out or is cancelled, an error is returned.
+func (s *solver) Solve(ctx context.Context) (result []Installable, err error) {
+	defer func() {
+		// This likely indicates a bug, so discard whatever
+		// return values were produced.
+		if derr := s.dict.Error(); derr != nil {
+			result = nil
+			err = derr
+		}
+	}()
+
+	g := gini.New()
+	s.dict.ApplyConstraints(g)
+
+	h := searcher{
+		solver:  s,
+		g:       g,
+		assumed: &orderedLitSet{indices: make(map[z.Lit]int)},
+	}
+	for _, anchor := range s.dict.MandatoryIdentifiers() {
+		h.assumed.Add(s.dict.LitOf(anchor))
+	}
+	g.Assume(h.assumed.Slice()...)
+
+	outcome, _ := g.Test(nil)
+	if outcome != satisfiable && outcome != unsatisfiable {
+		outcome = h.search(ctx, 0)
+	}
+	switch outcome {
+	case satisfiable:
+		if err := s.dict.MinimizeCardinality(g); err != nil {
+			return nil, fmt.Errorf("unexpected error: %w", err)
+		}
+		return s.dict.Installables(g), nil
+	case unsatisfiable:
+		return nil, NotSatisfiable(s.dict.Conflicts(g))
+	}
+
+	return nil, Incomplete
+}
+
+type orderedLitSet struct {
+	indices map[z.Lit]int
+	lits    []z.Lit
+}
+
+func (set *orderedLitSet) Add(m z.Lit) {
+	if set.Contains(m) {
+		return
+	}
+	set.indices[m] = len(set.lits)
+	set.lits = append(set.lits, m)
+}
+
+func (set *orderedLitSet) Remove(m z.Lit) {
+	if index, ok := set.indices[m]; ok {
+		set.lits = append(set.lits[:index], set.lits[index+1:]...)
+		delete(set.indices, m)
+	}
+}
+
+func (set *orderedLitSet) Contains(m z.Lit) bool {
+	_, ok := set.indices[m]
+	return ok
+}
+
+func (set *orderedLitSet) Slice() []z.Lit {
+	return set.lits
+}
+
+type searcher struct {
+	solver  *solver
+	g       *gini.Gini
+	assumed *orderedLitSet
+	buffer  []z.Lit
+}
+
+// ... Must always return either satisfiable or unsatisfiable!
+func (h *searcher) search(ctx context.Context, upto int) int {
+	if upto < len(h.assumed.Slice()) {
```

Comment:
nit: check if `>` and return early?

Reply:
I went back and forth on it, because I felt that it was clearest to end with `Solve()`. PTAL, I tried something different in my next push.